### ISSUE TITLE
Imports: Support imports from IPFS hashes

### ISSUE THREAD

**Issue by legowerewolf** [denoland/deno#5290]
IPFS as a source for dependencies would prevent possible security issues caused by hacked module sources. Files from IPFS are identified by their hash, and so it's impossible to modify a file without changing the file identifier. There are a number of public IPFS gateways available that make the content on the IPFS network available over standard HTTP(S), so the built-in URL-based module resolution would support IPFS imports from the gateways, but built-in access to the network could be an improvement. The IPFS community seems to be standardizing on `ipfs://` as a protocol identifier, which could be used to identify IPFS-based imports.

(More info and references and clarification to follow, but I'm on mobile at the moment.)

**Comment by joshverd** [comment]
How would you be able to identify which package you are importing if it's just a hash in the import statement?

**Comment by legowerewolf** [comment]
Doc comments on your `deps.ts` file? Ideally, the imported module would also have a doc comment at the top identifying itself.

**Comment by carsonfarmer** [comment]
The other option is sticking to a convention? You can reference linked data within an IPFS IPLD DAG structure (the stuff behind the hash) by name. This requires 'wrapping' the IPFS content in an external folder, but this isn't particularly burdensome.

```
import { Something } from "http://127.0.0.1:5001/ipfs/{hash}/some_package/mod.ts"
```

**Comment by legowerewolf** [comment]
That also works, and preserves the core immutability features we're looking for. Although, I think that import string would be better as 

```
import { Something } from "ipfs://{hash}/some_package/mod.ts"
```

to preserve the fallback options I proposed above.

**Comment by olanod** [comment]
Was going to create the exact same issue :sweat_smile: 
Another protocol that might be suitable as well is `p2p://`. I think this feature would be used a lot in tandem with [import maps](https://deno.land/manual/linking_to_external_code/import_maps), later the community could create cli tools to manage the json file more easily making use of some registry that maps package names to hashes, so you could do `module-management-tool add library` and it automatically adds `library` with the right hash.
```json
{
   "imports": {
      "library/": "p2p://<hash>"
   }
}
```
```js
import foo from 'library/foo.js'
```
And to go the extra mile that registry can just be another json file on ipfs managed by a decentralized blockchain based organization that is economically incentivized to keep a high quality registry with possibly different channels like `testing` and `stable` where packages are reviewed carefully to make sure no funny scripts make their way to the stable registry.

**Comment by Diomas** [comment]
IPFS is built on https://libp2p.io/implementations/ - it has both node.js and in-browser support.
So Deno just need to support that lib, and it can do bunch of p2p interactions then

**Comment by rotemdan** [comment]
From a security persepective, IPFS is not completely necessary, nor does it actually solve the issue of content-defined library imports - since it provides no guarantee that that the requested file would internally use a secure hash scheme to import its own dependencies, and so would their own dependencies' dependencies, and so on.

I think it would would also be productive to consider a scheme that is independent of any specific protocol (e.g. `ipfs`/`p2p`/`magnet`) and instead:

* Standardize some way to integrate the hash into the file's name, e.g.:
```
protocol://path/some_package@version/module.{hash}.ts
```
* Have Deno try to automatically verify the hash, including the hashes of all of the file's internal imports (if provided).
* Command-line option for a `strict` secure mode in which **every** imported file must include a hash.

For compatibility and future proofing: it seems reasonable to use [IPFS's multiformat scheme](https://cid.ipfs.io/) that allows to support various kinds of hash algorithms.

**Comment by rotemdan** [comment]
After thinking about this further I've found a simple alternative solution that guarantees 100% content addressed safety and reproducibility for the entire dependency tree:

It would look like this:
```
protocol://domain/path/name@version/module.ts#lock_cid=QmcRD4wkPPi6dig81r5sLj9Zm1gDCL4zgpEj9CfuRrGbzF
```

Where `lock_cid` is a content identifier (basically a hash in a flexible format) for a lock file that would include the hash of the target file (here `module.ts`) as well as the hashes of **all** imported files in the entire dependency tree (possibly including imports from the standard library):

The file might look like something like this (or might use JSON etc. this is only shown for simplicity):
```
https://my.website/path/name@version/module.ts Qmf8obm7bxrQS1JnjUniJdibcN2kUJy9zz732sr7o3dxtn
https://my.website/path/name@version/utils.ts Qmeg1Hqu2Dxf35TxDg18b7StQTMwjCqhWigm8ANgm8wA3p
https://my.website/path/name@version/methods.ts QmZfSNpHVzTNi9gezLcgq64Wbj1xhwi9wk4AxYyxMZgtCG

https://someother.website/path/name@version/othermodule.ts QmbKxNNCxBox7Cmv3jiUZbiG3zpzmtnYzVUuKHxfAjvpyH
https://someother.website/path/name@version/othermoduleutils.ts QmPwwoytFU3gZYk5tSppumxaGbHymMUgHsSvrBdQH69XRx

https://deno.land/std@v0.3.0/async/delay.ts QmaLRet8qeYqNaq8xJeiqwjNnukSo3uEA8oWsDLoxxBv4Q
https://deno.land/std@v0.3.0/async/deferred.ts QmWZtn3ahqqpGBBRZqPdthcWz2n1rxc1UuiDoWXrgrHKzZ

...

```

Since the lock file is content addressed, it can be fetched from anywhere, either from IPFS, or from the web server itself (say in `https://my.website/path/name@version/.deno-lock`).

The lock file can also be effectively used as an IPFS-like merkledag (though unlike in IPFS, it doesn't represent a directory structure, but a collection of references to various sources), but since all the references use `cid`s, they can all be potentially fetched from IPFS (and in parallel, which may also improve performance).

Technically, if some of the imports in the dependency tree already refer to their own lock file, then it may not be strictly necessary to include them in the lock file. However, since the storage requirements of a structure like this are relatively minimal in modern standards, it may be better not to rely on multiple sources and include everything in one file (even if techincally redundant).

**Comment by srdjan** [comment]
ha! I had similar thoughts here: https://github.com/denoland/deno_website2/issues/406#issuecomment-633298685

**Comment by rotemdan** [comment]
@srdjan 
Interesting you were going the same way.

I've modified this approach through several iterations and eventually came to a solution that doesn't actually require the library authors to know anything about the existance of a lock file or even the hashing scheme. I'll try to clarify some aspect for the design :

The hash part (`#`) is never actually sent to the server, it is only for local use:
```
https://domain/path/name@version/module.ts#lock_cid=QmcRD4wkPPi6dig81r5sLj9Zm1gDCL4zgpEj9CfuRrGbzF
```
The lock file is individual to the dependecy tree of a particular `ts` file (there is no consideration of a directory structure here), but by convention, for the above URI, **one** of its search location might be:
```
https://domain/path/name@version/module.ts.lock
```
However, it doesn't have to. It can be stored locally or fetched from a p2p network like IPFS.

As an importer of a third-party `ts` or `js` file, you would be able to produce, **by yourself**, a lock file for that particular import and put it practically wherever you want.

**Comment by rotemdan** [comment]
Alright, after considering it even more, I've realized that the lock file isn't even strictly necessary to be stored anywhere. Instead, it can be regenerated purely based the content and structure of the dependency tree for verification purposes.

It's actually pretty simple:

Say I want to import this URI, but I also want a strong proof that ensures that what I get is always the same:
```js
import * from "https://example.com/path/name@version/module.ts"
```

I use a tool that walks the dependency tree of `module.ts` in some deterministic order and records the hashes of all the files it finds and puts it in some file called `module.ts.lock`. I annotate the hash of that file into the link like this:
```js
import * from "https://example.com/path/name@version/module.ts#lock_cid=QmaLRet8qeYqNaq8xJeiqwjNnukSo3uEA8oWsDLoxxBv4Q"
```

Now whenever someone encounters this annotated link. They have two options:
* Try to fetch the lock file from somewhere (server, local directory, IPFS) and use it for verification.
* Recreate the same process (regenerate the lock file) and see if the hash they get is the same.

That's pretty much it.

**Comment by legowerewolf** [comment]
As an update, [this happened.](https://www.chromestatus.com/feature/4776602869170176) Chrome extended support for custom protocol handlers to include, among others, support for `ipfs://` and `ipns://`.

**Comment by RobotSail** [comment]
Going off of some comments that have been posed in this thread, would it make sense for there to be a convention that utilizes IPNS as part of its base path to host modules? 
Since the base ID for an IPNS directory is just the multihash of a public key, this could be used to create signatures for the modules to verify the signing authority. 

So someone responsible for creating a module could host the directory at `ipns/<foo_pubkey_hash>` with the following structure:
```
pubkey.txt
modules/
    foo_module/
        index.ts
        index.ts.sig
    bar_module/
        index.ts
        index.ts.sig
```

Then the directory could be declared from within a `package.json`, and the necessary import logic could be handled by the runtime:
```json
{
    "imports": {
        "alias_name": {
            "cid": "ipns/<pubkey_hash>",
            "modules": ["foo_module", "bar_module"] 
        }
}
```

The files themselves could then be imported as such: 
```ts
import React from "ipfs:alias_name/foo_module"
```

The runtime could check the signatures of the modules being requested for import and error if they don't line up, or if the pubkey doesn't match its IPNS hash.

**Comment by jeff-hykin** [comment]
For me, it seems like IPNS would kind of defeat the purpose of IPFS for Deno.

I'd like the hash to be in the import statement, and the URL to always point to the same content, instead of needing a separate package.json file to map. At that point, it just seems like a package lock file to a regular website whose content can change to be anything.

**Comment by locriacyber** [comment]
IPFS uses multihash, which is incompatible with URL.

**Comment by RangerMauve** [comment]
Just wanted to chime in that it would be nice if we could have a way to add new protocol schemes for imports (or maybe for fetch as well from within code).

I'm the main developer of [Agregore](https://agregore.mauve.moe/), a p2p web browser that supports stuff like IPFS. I'd love it if we could reuse modules published for p2p web browsers like Agregore within Deno.

In Agregore, we've extended the browser's protocol handlers to support [ipfs/ipns/etc](https://github.com/RangerMauve/js-ipfs-fetch/) not just for reading with GET, but also for writing with PUT.

On desktop we're doing this via the [protocol](https://www.electronjs.org/docs/latest/api/protocol/) API provided by Electron, and via some C++ changes [on mobile](https://github.com/AgregoreWeb/agregore-mobile/blob/default/patches/0001-AG-Handle-IPFS-and-IPNS-URLs.patch).

Similarly, the [Node.js VM API](https://nodejs.org/api/vm.html) enables us to provide custom "linkers" which can enable customizing module resolution for custom protocols (something I've played around with in [webrun](https://github.com/RangerMauve/webrun)).

Having a way to dynamically add protocol scheme handlers in Deno would make it easier to integrate stuff like IPFS in user-land without needing large dependencies.

**Comment by legowerewolf** [comment]
Actually, yeah, I think that would be a good way to go about it. If you could register new protocol schemes for Deno (probably with a `Deno`-namespaced function), you could support any protocol you wanted. If this registration also added support for those protocols to the `fetch` function, it'd be even more useful.

**Comment by legowerewolf** [comment]
Yeah. It'd be good if it was built-in. It might not even be that hard - has anyone tried running js-ipfs inside Deno? It might *just work*. Sure, implementing support for talking to a user's existing node would be good, but I'm pretty sure the JS version is stable enough for our needs here.

**Comment by rafrafek** [comment]
In Python when installing dependencies you can specify hashes:

```
requests==2.31.0 ; python_full_version >= "3.12.0" and python_full_version < "3.13.0" \
    --hash=sha256:58cd2187c01e70e6e26505bca751777aa9f2ee0b7f4300988b709f44e013003f \
    --hash=sha256:942c5a758f98d790eaed1a29cb6eefc7ffb0d1cf7af05c3d2791656dbd6ad1e1
```

Poetry by default generates lock file with hashes and if you export dependencies to a requirements.txt file it will include hashes as well.

That way you can be sure nobody tempered with the content of your dependency module.

**Comment by lucacasonato** [comment]
We are not going to do this.
